zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/distribution/DbDistributionState.java

```diff
@@ -42,6 +44,16 @@ public class DbDistributionState implements MutableDistributionState {
   private final ColumnFamily<DbLong, PersistedCommandDistribution>
       commandDistributionRecordColumnFamily;
 
+  /** [queue | partition id | insertion key ] => [distributionKey] */
+  private final ColumnFamily<
+          DbCompositeKey<DbString, DbCompositeKey<DbInt, DbLong>>, DbForeignKey<DbLong>>
+      queuedCommandDistributionColumnFamily;
```

Comment:
I'm not sure if this is correct. The `distributionKey` is not guaranteed to be a freshly generated key, it can also just be some old key that would then mess up the queue order, right?

Reply:
See for example the `ProcessInstanceMigrationMigrateProcessor` which uses the message subscription key directly as distribution key

Reply:
In my opinion that's incredibly dirty and should be changed in this processor. 

@berkaycanbc Do you know why we use the message subscription key here? I see in the `MessageSubscriptionMigrateProcessor` it's only used for some rejection message. Why don't we generate a new distribution key?

Reply:
To clarify: I've introduced the `insertionKey` not for uniqueness but to maintain FIFO ordering, under the assumption that `distributionKey` is _not_ required to be generated freshly when starting the distribution.
If we now decide that this should be part of the overall contract then we should find a way to enforce that. For example by not accepting `distributionKey` from the outside but instead generating it internally. If we do that, we could remove the `insertionKey` again because `distributionKey` already guarantees the queue order.

Reply:
@berkaycanbc Fyi we found it explained [here](https://github.com/camunda/camunda/pull/18822/commits/dfb3a7c986135ac6ac1da9c5769224f8de6fdd1b)

>1) The order of records arriving to the other partition via command distribution cannot be
guaranteed. For example, when two commands are distributed after each other, then any of them
could fail and be retried. Meaning that, the later one can arrive first successfully.
> 
> This is problematic for migrations because we might have same process instance migrated twice and
the message subscription migration of the second migration could arrive at the related partition
first. Because that reason, we used process message subscription key as a distribution key which
must be unique until the distribution is completed. Then, we will check if there are any pending
distributions of a message subscription and reject following migration until the distribution is
finished.

So it has to do with non-deterministic ordering of command distributions. If we have this queue solution are you good with generating a fresh distribution key? It's more fail-safe than the current solution anyway 😄

Reply:
I was impatient and already pushed a commit removing the insertion key and relying on distribution key order instead. We can roll back this commit if we decide against it. 8d121ff

Reply:
@remcowesterhoud @lenaschoenburg 
> So it has to do with non-deterministic ordering of command distributions. If we have this queue solution are you good with generating a fresh distribution key?

Exactly. Using the `queueId` in `ProcessInstanceMigrationMigrateProcessor` would serve the same purpose. Using `subscriptionKey` as the distribution key was the trick we implemented to prevent disordered message subscription migration. E.g. if 2 migrations triggered after another and both of them changes the message name expression (not supported yet but will be supported). If the second msg sub migration command arrives the partition first, the first msg sub migration command will override the message name incorrectly.

> It's more fail-safe than the current solution anyway 😄

Not sure if this is fail-safe but in that case we accept the fact that the command distribution is fire and forget. If something goes wrong with migrating msg sub we still do not know and mark the migration as successful.

Reply:
Btw, I'll add a way to register custom continuation commands for command distribution this week: https://github.com/camunda/camunda/issues/21449. Maybe that helps in that case too because you can await successful distribution.

Reply:
@lenaschoenburg that would improve the UX. E.g. do not complete migration if the msg sub distribution is not completed. But it also depends a bit on how continuation will be implemented.

Reply:
I'd be super happy if you could leave some feedback on https://github.com/camunda/camunda/issues/21449 then, e.g. describing what you'd need out of it, how it should work etc.
I have something in mind but I'd be happy to adjust the implementation so that it is useful for others too.

Reply:
IMO we should use queueId for all distributions related to message subscriptions. This will also prevent any update msg sub (or process msg sub) state to arrive in order. E.g. a msg sub is updated after the migration but update command processed first.

Reply:
Created an issue https://github.com/camunda/camunda/issues/21969 to use `queueId` for msg sub migration.